bugtool/cmd/root.go

```diff
@@ -0,0 +1,354 @@
+	// commands to use `kubectl`
+	k8sPods, err := getCiliumPods(k8sNamespace, k8sLabel)
+	if err != nil { // Assuming this is a non k8s deployment
+		dst := filepath.Join(ciliumDir, defaults.StateDir)
+		if err := copyDir(stateDir, dst); err != nil {
+			log.WithError(err).Warn("could not copy state directory")
+		}
+
+		for _, cmd := range sources {
+			// Add the host flag if set
+			if len(host) > 0 {
+				cmd = fmt.Sprintf("%s -H %s", cmd, host)
+			}
+			writeCmdToFile(ciliumDir, cmd)
+		}
+	} else { // Found k8s pods
+		for _, pod := range k8sPods {
+			dst := filepath.Join(ciliumDir, fmt.Sprintf("%s-%s", k8sPods, defaults.StateDir))
+			kubectlArg := fmt.Sprintf("%s/%s:%s", k8sNamespace, pod, stateDir)
+			// kubectl cp kube-system/cilium-xrzwr:/var/run/cilium/state cilium-xrzwr-state
+			_, err := execCommand("kubectl", "cp", kubectlArg, dst)
+			if err != nil {
+				log.WithError(err).Warn("could not copy state directory")
+			}
+
+			prefix := fmt.Sprintf("kubectl exec %s", pod)
+			for _, cmd := range sources {
+				// Add the host flag if set
+				if len(host) > 0 {
+					cmd = fmt.Sprintf("%s%s -H %s", prefix, cmd, host)
+				}
+				writeCmdToFile(ciliumDir, cmd)
+			}
+		}
+	}
+}
+
+func execCommand(cmd string, args ...string) (string, error) {
+	log.Printf("exec: %s %s", cmd, args)
+	output, err := exec.Command(cmd, args...).CombinedOutput()
+	return string(output), err
+}
+
+// writeCmdToFile will execute command and write markdown output to a file
+func writeCmdToFile(cmdDir, prompt string) {
+	// Clean up the filename
+	name := strings.Replace(prompt, "/", " ", -1)
+	name = strings.Replace(name, " ", "-", -1)
+	f, err := os.Create(filepath.Join(cmdDir, name+".md"))
+	if err != nil {
+		log.WithError(err).Warnf("Could not create file")
+		return
+	}
+	defer f.Close()
+
+	// Split the command and arguments
+	split := strings.Split(prompt, " ")
+	argc := len(split)
+	var args []string
+	cmd := split[0]
+
+	// The command does not exist, abort.
+	if _, err := exec.LookPath(cmd); err != nil {
+		os.Remove(f.Name())
+		return
+	}
+
+	if argc > 1 {
+		args = split[1:]
+	}
+
+	// Write prompt as header and the output as body, and / or error but delete empty output.
+	output, err := execCommand(cmd, args...)
+	if err != nil {
+		fmt.Fprintf(f, fmt.Sprintf("> Error while running '%s':  %s\n\n", prompt, err))
+	}
+	// We deliberately continue in case there was a error but the output
+	// produced might have useful information
+	if strings.Contains(output, "```") {
+		// Already contains Markdown, print as is.
+		fmt.Fprint(f, output)
+	} else if len(output) > 0 {
+		fmt.Fprintf(f, fmt.Sprintf("# %s\n\n```\n%s\n```\n", prompt, output))
+	} else {
+		// Empty file
+		os.Remove(f.Name())
+	}
+}
+
+func getCiliumPods(namespace, label string) ([]string, error) {
+	output, err := execCommand("kubectl", "-n", namespace, "get", "pods", "-l", label)
+	ciliumPods := []string{}
+	if err != nil {
+		return ciliumPods, err
```

Comment:
you could return the literal directly here (`return []string{}, err`) since it isn't used otherwise. Normally if a function returns an error it isn't expected that the other return variables are valid so you could also return a nil (`return nil, err`). It just so happens that nil is an empty slice so most operations (like len, range and cap) will return reasonable values, so it isn't as bad as a null would be in C.

Reply:
Sounds good.